### PR TITLE
Add _container.task.framework.mesos A records with container IP

### PR DESCRIPTION
**Pull request by sttts**
For each existing A record for a task with a container IP another A record is added with the `_container` prefix, resolving to the container IP instead of the host IP.

The container IP is passed in a status label of a task in state.json, e.g. 

```json
{
  "executor_id": "",
  "framework_id": "20140703-014514-3041283216-5050-5348-0000",
  "id": "liquor-store.b8db9f73-562f-11e4-a088-c20493233aa5",
  "name": "liquor-store",
  "slave_id": "20140803-125133-3041283216-5050-2410-0",
  "state": "TASK_RUNNING",
  "statuses": [
    {
      "state": "TASK_RUNNING",
      "timestamp": 1413572390.09275,
      "labels": [
        {
          "key": "Docker.NetworkSettings.IPAddress",
          "value": "10.3.0.1"
        }
      ]
    }
  ]
}
```

Compare https://reviews.apache.org/r/36537/ and https://issues.apache.org/jira/browse/MESOS-3076

### CODE REVIEW

File: records/config.go [hunk 1]
```diff
@@ -71,6 +71,9 @@ type Config struct {
 
 	// EnforceRFC952 will enforce an older, more strict set of rules for DNS labels
 	EnforceRFC952 bool
+
+	// parse docker inspect output for container ip
```
Comment (tsenart) [inline records/config.go]: Comment should begin with the described variable.

File: records/generator.go [hunk 2]
```diff
@@ -36,13 +37,21 @@ type Resources struct {
 	Ports string `json:"ports"`
 }
 
+// Status hold a task status
```
Comment (tsenart) [inline records/generator.go]: s/hold/represents/

File: records/generator.go [hunk 3]
```diff
@@ -36,13 +37,21 @@ type Resources struct {
...
+}
+
 // Tasks holds mesos task information read in from state.json
 type Tasks []struct {
-	FrameworkId string `json:"framework_id"`
-	Id          string `json:"id"`
-	Name        string `json:"name"`
-	SlaveId     string `json:"slave_id"`
-	State       string `json:"state"`
+	FrameworkId string   `json:"framework_id"`
```
Comment (tsenart) [inline records/generator.go]: s/Id/ID/

File: records/generator.go [hunk 4]
```diff
@@ -36,13 +37,21 @@ type Resources struct {
...
+	Id          string   `json:"id"`
+	Name        string   `json:"name"`
+	SlaveId     string   `json:"slave_id"`
```
Comment (tsenart) [inline records/generator.go]: s/Id/ID/

File: records/generator.go [hunk 5]
```diff
@@ -66,6 +75,16 @@ type StateJSON struct {
 	Leader     string `json:"leader"`
 }
 
+// DockerNetworkSettings partially holds the docker container network settings
+type DockerNetworkSettings struct {
+	IPAddress string `json"IPAddress"`
```
Comment (tsenart) [inline records/generator.go]: Typo in the field tag. Also, we're using snake casing for all other json key names. We should be consistent.

File: records/generator.go (hunk 3 above)
Comment (sttts) [inline records/generator.go]: not part of this commit

File: records/generator.go [hunk 6]
```diff
@@ -248,12 +268,38 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
...
+						}
+
+						if len(inspectResult) > 0 {
+							containerIP = inspectResult[0].NetworkSettings.IPAddress
+						}
+						break
+					}
+				}
+			}
+
```
Comment (tsenart) [inline records/generator.go]: Can we extract this to a private method? `InsertState` is getting quite long.

File: records/generator.go [hunk 7]
```diff
@@ -248,12 +268,38 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
...
 			// skip not running or not discoverable tasks
 			if !ok || (task.State != "TASK_RUNNING") {
 				continue
 			}
 
+			// parse docker inspect output for container IP if enabled
+			containerIP := ""
+			if useDockerContainerIP {
+				for _, status := range task.Statuses {
+					if status.State == "TASK_RUNNING" && len(status.Data) > 0 {
```
Comment (tsenart) [inline records/generator.go]: Reduce nesting by inverting the condition and continuing.

File: records/generator.go [hunk 8]
```diff
@@ -248,12 +268,38 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
...
+						err = json.Unmarshal(decoded, &inspectResult)
+						if err != nil {
+							logging.VeryVerbose.Printf("error parsing docker inspect output of task %v: %v", task.Id, err)
+							continue
+						}
+
+						if len(inspectResult) > 0 {
+							containerIP = inspectResult[0].NetworkSettings.IPAddress
+						}
+						break
```
Comment (tsenart) [inline records/generator.go]: I would think that you'd only want to break if the containerIP was actually found. Here we're breaking anyways.

File: records/generator_test.go [hunk 9]
```diff
@@ -256,10 +255,18 @@ func TestInsertState(t *testing.T) {
 		t.Error("should not find this not-running task - SRV record")
 	}
 
-	_, ok = rg.As["liquor-store.marathon.mesos."]
+	As, ok := rg.As["liquor-store.marathon.mesos."]
```
Comment (tsenart) [inline records/generator_test.go]: It's kind of weird to use capitalized names for local variables.

File: records/generator_test.go [hunk 10]
```diff
@@ -256,10 +255,18 @@ func TestInsertState(t *testing.T) {
...
 	if !ok {
 		t.Error("should find this running task - A record")
 	}
+	if len(As) != 2 || As[0] != "1.2.3.11" || As[1] != "1.2.3.12" {
```
Comment (tsenart) [inline records/generator_test.go]: Easier: `if got, want := As, []string{"1.2.3.11", "1.2.3.12"}; !reflect.DeepEqual(got, want) {`

File: records/generator_test.go (hunk 9 above)
Comment (sttts) [inline records/generator_test.go]: yes, but here it is A like A-record. kind of makes sense

Comment (tsenart) [inline records/generator_test.go]: I'd just call it `rrs` for resource records.

**Comment by tsenart** [comment]
@sttts: PTAL

File: records/generator_test.go (hunk 9 above)
Comment (sttts) [inline records/generator_test.go]: there are also SRV records

File: records/generator_test.go (hunk 10 above)
Comment (sttts) [inline records/generator_test.go]: well, easier can be discussed.

Comment (sttts) [inline records/generator_test.go]: in fact I would prefer a set equal

Comment (tsenart) [inline records/generator_test.go]: Set equal?

File: records/generator_test.go (hunk 9 above)
Comment (sttts) [inline records/generator_test.go]: ips makes sense

Comment (sttts) [inline records/generator_test.go]: but rrs is used elsewhere already

File: records/generator_test.go (hunk 10 above)
Comment (sttts) [inline records/generator_test.go]: like set(list) = set(list2)

File: records/generator.go [hunk 11]
```diff
@@ -231,14 +250,42 @@ func sanitizedSlaveAddress(hostname string, spec labels.HostNameSpec) string {
 	return address
 }
 
+func (task *Task) containerIP() string {
+	for _, status := range task.Statuses {
+		if status.State != "TASK_RUNNING" || len(status.Data) == 0 {
+			continue
+		}
+
+		decoded, err := base64.StdEncoding.DecodeString(status.Data)
```
Comment (tsenart) [inline records/generator.go]:
```go
var data []DockerInspect
dec := json.NewDecoder(base64.NewDecoder(base64.StdEncoding, bytes.NewReader(status.Data))
if err := dec.Decode(data); err != nil {
    // log error
} else if len(data) > 0 {
    return data[0].NetworkSettings.IpAddress
}
```

```

Comment (sttts) [inline records/generator.go]: Why not using the StdEncoding.DecodeString?

File: records/generator_test.go [hunk 12]
```diff
@@ -317,6 +324,48 @@ func TestInsertState(t *testing.T) {
...
+	}
+	if got, want := rrs, []string{"1.2.3.11", "1.2.3.12"}; !reflect.DeepEqual(got, want) {
+		t.Errorf("should return the slave ips 1.2.3.11, 1.2.3.12 for the task, but got %v", rrs)
+	}
+
+	rrs, ok = rg.As["_container.liquor-store.marathon.mesos."]
+	if !ok {
+		t.Error("should find the container ip because ParseDockerInspectForContainerIP is enabled")
+	}
+	if len(rrs) != 1 || rrs[0] != "10.3.0.1" {
```
Comment (tsenart) [inline records/generator_test.go]: If we're now using `reflect.DeepEqual` in the previous case, we should use it in the others too.

File: records/generator.go (hunk 11 above)
Comment (tsenart) [inline records/generator.go]: Composition FTW!

File: factories/fake.json [hunk 13]
```diff
@@ -206,7 +212,49 @@
...
-                            "timestamp": 1413572390.09275
+                            "timestamp": 1413572371.04321,
+                            "labels": [
+                                {
+                                    "key": "foo",
+                                    "value": "bar"
+                                },
+                                {
+                                    "key": "docker_inspect",
+                                    "value": "[\n{\n \"NetworkSettings\": {\n \"GlobalIPv6PrefixLen\": 0,\n \"HairpinMode\": false,\n \"IPAddress\": \"10.3.0.4\" }\n}\n]\n"
```
Comment (tsenart) [inline factories/fake.json]: Are the newlines inside the JSON needed?

File: records/generator.go [hunk 14]
```diff
@@ -17,6 +17,10 @@ import (
 	"github.com/mesosphere/mesos-dns/records/labels"
 )
 
+const (
+	DockerInspectLabel = "docker_inspect"
```
Comment (tsenart) [inline records/generator.go]: Please add a comment describing this constant.

File: factories/fake.json (hunk 13 above)
Comment (sttts) [inline factories/fake.json]: It's a real extract of a state.json.

File: records/generator.go (hunk 14 above)
Comment (sttts) [inline records/generator.go]: done.

File: records/generator.go [hunk 15]
```diff
@@ -66,6 +84,16 @@ type StateJSON struct {
...
+// DockerNetworkSettings is a partial representation of the docker container network settings
+type DockerNetworkSettings struct {
+	IpAddress string `json:"IpAddress"`
+}
+
+// DockerInspect is a partial representation of a docker inspect output for a task's container
+type DockerInspect struct {
+	NetworkSettings DockerNetworkSettings `json:"NetworkSettings"`
+}
+
```
Comment (tsenart) [inline records/generator.go]: We probably don't need to export these types. It'd be enough to declare them only in the scope they're required, inside `containerIPFromDockerInspect`.
```go
func containerIPFromDockerInspect(inspect string) (string, error) {
    var data []struct {
        Settings struct {
            IP string `json:"IpAddress"`
        } `json:"NetworkSettings"`
    }
    // ...
}
```

File: records/generator.go [hunk 16]
```diff
@@ -231,14 +252,66 @@ func sanitizedSlaveAddress(hostname string, spec labels.HostNameSpec) string {
...
 }
 
+func containerIPFromDockerInspect(inspect string) (string, error) {
+	var inspectJSON []struct {
+		NetworkSettings struct {
+			IpAddress string `json:"IpAddress"`
+		} `json:"NetworkSettings"`
+	}
+
+	err := json.Unmarshal([]byte(inspect), &inspectJSON)
```
Comment (tsenart) [inline records/generator.go]: `inspectJSON` is a reference type so no need to pass a reference in to `json.Unmarshal`

**Comment by tsenart** [comment]
LGTM

**Comment by tsenart** [comment]
@karya0: PTAL

File: records/generator.go [hunk 17]
```diff
@@ -17,6 +17,12 @@ import (
 	"github.com/mesosphere/mesos-dns/records/labels"
 )
 
+const (
+	// containerIPTaskStatusLabel represents the key of the state.json task status label
+	// which holds the container IP.
+	containerIPTaskStatusLabel = "Docker.NetworkSettings.IPAddress"
```
Comment (tsenart) [inline records/generator.go]: This isn't used anywhere else than the `containerIP` method. We should define it inside and get rid of the comment since its not necessary so close to where it's used.
